### PR TITLE
Chat: add transcript messages in pairs

### PR DESCRIPTION
**Pull request by abeatrix**
FIX: https://github.com/sourcegraph/cody/issues/3126

Update to remove consecutive assistant messages after a human message cannot be added due to context limit reach.

Currently, we would stop adding old transcript messages when the limit has been reached, which was causing `found consecutive messages with the same speaker 'assistant'` error because we are not removing the (assistant) response that is associated with the (human) message we are skipping (as we loop through them from back to front).

~~This PR updates to removing the last assistant message responding to the current human message when the limit has been reached, instead of keeping assistant response without adding human message. This prevents the error when adding the preamble ending in 'assistant'.~~

#### Updated

Changes included:
- removed `tryAdd` which was adding message individually
- added `tryAddMessages` which takes the reversed transcript before adding the messages in pairs (human and assistant messages). When the limit has been reached, both the human and assistant messages will not be added. 
  - also check to ensure an assistant message must be followed by a human message
- added unit test for `tryAddMessages`
- removed the unused `CommandPrompter`

## Test plan

Keep asking Cody questions until you have reached the context limit (check output channel for something similar to `DefaultPrompter.makePrompt: Ignored 19 transcript messages due to context limit`). You should not encounter the `found consecutive messages with the same speaker 'assistant'` error.

#### Before

Screenshot 2024-02-20 at 1 24 30 PM

Output Channel - the first message after the added preamble is from `assistant`:
```
█ DefaultPrompter.makePrompt: Ignored 19 transcript messages due to context limit
█ CompletionLogger:onError: {"type":"completion","endpoint":"https://sourcegraph.com/.api/completions/stream","status":"error","duration":302,"err":"found consecutive messages with the same speaker 'assistant'"}  {
  "params": {
    "temperature": 0.2,
    "maxTokensToSample": 1000,
    "topK": -1,
    "topP": -1,
    "model": "anthropic/claude-2.0",
    "messages": [
      {
        "speaker": "human",
        "text": "You are Cody, an AI coding assistant from Sourcegraph."
      },
      {
        "speaker": "assistant",
        "text": "I am Cody, an AI coding assistant from Sourcegraph."
      },
      {
        "speaker": "assistant",
        "text": " Based on reviewing the terminal output:\n\n- It is building a VS Code extension using pnpm, Vite, esbuild and referencing @sourcegraph packages.\n\n- There are deprecation warnings about Vite's CJS API that should be addressed by upgrading Vite.\n\n- The main output files are extension.node.js (10.7mb) and extension.node.js.map (18.9mb).\n\n- Build times are fast, around 300-400ms.\n\n- The build commands complete successfully with exit code 0.\n\nTo address the deprecation warnings:\n\n1. Open package.json and identify the Vite version.\n\n2. Update Vite to the latest stable version.\n\n3. Run `pnpm install` to install updated dependencies.\n\n4. Re-run `pnpm run build:dev:desktop` \n\n5. The warnings should no longer occur.\n\nLet me know if you need any clarification or have additional questions!"
      },
      {
        "speaker": "human",
        "text": "\nReview and analyze this terminal output from the `node` process and summarize the key information. If this indicates an error, provide step-by-step instructions on how I can resolve this:\n\n```\n\n*  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 368ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 349ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 405ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 337ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 426ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 376ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it.\n\n```\n\nProcess options: {\"name\":\"Build VS Code Extension (Desktop)\",\"shellPath\":\"/bin/zsh\",\"shellArgs\":[\"-l\",\"-c\",\"pnpm --silent run build:dev:desktop\"],\"cwd\":{\"$mid\":1,\"fsPath\":\"/Users/bwork/dev/cody-3/vscode\",\"path\":\"/Users/bwork/dev/cody-3/vscode\",\"scheme\":\"file\"},\"env\":{}}"
      },
      {
        "speaker": "assistant",
        "text": " Here is a summary of the key information from the terminal output:\n\n- This is building a VS Code extension using pnpm, Vite, esbuild and @sourcegraph packages.\n\n- There are deprecation warnings for the Vite CJS API that should be addressed. \n\n- The main output files are extension.node.js (10.7mb) and extension.node.js.map (18.9mb).\n\n- Build times are fast - around 300-400ms.\n\n- Build commands complete successfully with exit code 0.\n\nTo fix the deprecation warnings:\n\n1. Open package.json and identify the current Vite version.\n\n2. Update Vite to the latest stable version.\n\n3. Run `pnpm install` to update dependencies.\n\n4. Re-run the build command: `pnpm run build:dev:desktop`\n\n5. The deprecation warnings should no longer occur after updating Vite.\n\nLet me know if you need any clarification or have additional questions!"
      },
```

#### After

Screenshot 2024-02-20 at 1 35 30 PM

Output Channel - the first message after the added preamble is from `human`, and the `assistant` message has been removed:
```
█ DefaultPrompter.makePrompt: Ignored 19 transcript messages due to context limit
█ logEvent (telemetry disabled): CodyVSCodeExtension:chat-question:executed VSCode {"properties":{"requestID":"33bda8b1-1a9a-45b3-81c3-97d59973c294","chatModel":"anthropic/claude-2.0","traceId":"c165b0facf0167c1398dc34f2ec413b0","contextSummary":{}},"opts":{"hasV2Event":true}}
█ telemetry-v2: recordEvent: cody.chat-question/executed: {"parameters":{"version":0,"metadata":[{"key":"recordsPrivateMetadataTranscript","value":1}],"privateMetadata":{"properties":{"requestID":"33bda8b1-1a9a-45b3-81c3-97d59973c294","chatModel":"anthropic/claude-2.0","traceId":"c165b0facf0167c1398dc34f2ec413b0","contextSummary":{}},"promptText":"hey"}},"timestamp":"2024-02-20T21:51:45.960Z"}
█ CompletionLogger:onComplete: {"type":"completion","endpoint":"https://sourcegraph.com/.api/completions/stream","status":"success","duration":9126}  {
  "result": " Hello!",
  "params": {
    "temperature": 0.2,
    "maxTokensToSample": 1000,
    "topK": -1,
    "topP": -1,
    "model": "anthropic/claude-2.0",
    "messages": [
      {
        "speaker": "human",
        "text": "You are Cody, an AI coding assistant from Sourcegraph."
      },
      {
        "speaker": "assistant",
        "text": "I am Cody, an AI coding assistant from Sourcegraph."
      },
      {
        "speaker": "human",
        "text": "\nReview and analyze this terminal output from the `node` process and summarize the key information. If this indicates an error, provide step-by-step instructions on how I can resolve this:\n\n```\n\n*  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 368ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 349ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 405ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 337ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 426ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it. \n\n *  Executing task: pnpm --silent run build:dev:desktop \n\n[1] The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.\n[0] All wasm modules are in place, have a good day!\n[0] \n[0]   dist/extension.node.js      10.7mb ⚠️\n[0]   dist/extension.node.js.map  18.9mb\n[0] \n[0] ⚡ Done in 376ms\n[0] pnpm run -s _build:esbuild:desktop --alias:@sourcegraph/cody-shared=@sourcegraph/cody-shared/src/index --alias:@sourcegraph/cody-shared/src=@sourcegraph/cody-shared/src --alias:@sourcegraph/cody-ui=@sourcegraph/cody-ui/src/index --alias:@sourcegraph/cody-ui/src=@sourcegraph/cody-ui/src exited with code 0\n[1] pnpm run -s _build:webviews --mode development exited with code 0\n *  Terminal will be reused by tasks, press any key to close it.\n\n```\n\nProcess options: {\"name\":\"Build VS Code Extension (Desktop)\",\"shellPath\":\"/bin/zsh\",\"shellArgs\":[\"-l\",\"-c\",\"pnpm --silent run build:dev:desktop\"],\"cwd\":{\"$mid\":1,\"fsPath\":\"/Users/bwork/dev/cody-3/vscode\",\"path\":\"/Users/bwork/dev/cody-3/vscode\",\"scheme\":\"file\"},\"env\":{}}"
      },
      {
        "speaker": "assistant",
        "text": " Here is a summary of the key information from the terminal output:\n\n- This is building a VS Code extension using pnpm, Vite, esbuild and @sourcegraph packages.\n\n- There are deprecation warnings for the Vite CJS API that should be addressed. \n\n- The main output files are extension.node.js (10.7mb) and extension.node.js.map (18.9mb).\n\n- Build times are fast - around 300-400ms.\n\n- Build commands complete successfully with exit code 0.\n\nTo fix the deprecation warnings:\n\n1. Open package.json and identify the current Vite version.\n\n2. Update Vite to the latest stable version.\n\n3. Run `pnpm install` to update dependencies.\n\n4. Re-run the build command: `pnpm run build:dev:desktop`\n\n5. The deprecation warnings should no longer occur after updating Vite.\n\nLet me know if you need any clarification or have additional questions!"
      },
```

### CODE REVIEW

**Review by beyang — CHANGES_REQUESTED** [review]
Does this fix the underlying error? This seems like it would mask an error that happens earlier?

**Comment by abeatrix** [comment]
> This seems like it would mask an error that happens earlier?

I don't think this is correct because we are handling the transcript in reverse, so we are currently adding the assistant response before we add the human question, which is causing the issue we are seeing because when the limit has been reached, and we skip adding the human question, the assistant response has already been added, which was causing the error message `found consecutive messages with the same speaker 'assistant'` to show up when we submit the `truncated` transcripts.

File: vscode/src/prompt-builder/index.ts [hunk 1]
```diff
@@ -34,6 +35,9 @@ export class PromptBuilder {
         return true
     }
 
+    /**
+     * @deprecated Use 'tryAddMessages' instead.
+     */
```
Comment (beyang) [inline vscode/src/prompt-builder/index.ts:40]: Why not just remove? It's unused.

**Review by beyang — APPROVED** [review]

File: vscode/src/prompt-builder/index.ts (hunk 1 above)
Comment (abeatrix) [reply vscode/src/prompt-builder/index.ts]: Removed. I wasn't sure if it would affect other clients. Thanks!

**Pull request merged**